### DISCUSSION TITLE
How to search for newline character

### DISCUSSION THREAD

**Discussion by a1ess** [orbitalquark/textadept#86] [General]
Hi,

I've been trying to match a newline character (\n) using the Search dialog but to no avail: I've tried both the normal search and the Regex search with \n and always get No results found. I'm on Linux with Textadept v. 11.1.

Is there a special syntax to match newlines and other escape sequences in Textadept?

Thanks.

**Comment by orbitalquark** [comment]
With regex enabled, try '$'. It's supposed to match the EOL.

**Reply by a1ess** [reply]
Thanks! '$' actually works, but only for the first occurrence. In a text like the following:

one
two
three

If the cursor is before the 'o' of 'one' and I search for $, it will match the newline character on that line, but won't find any more matches with "Find Next". Also, that only match cannot be replaced by anything using the Replace field.

My goal was actually to convert the lines (each newline character) into a comma, so that I can have all the elements in one line; something like:

Find : \n 
Replace: , 

I could probably write a script to achieve that. My  knowledge of Lua is pretty limited, but for what I've seen, Lua has no built-in functions to split/join strings, so I would have to write my own.

Just wanted to check if this kind of search/replace was possible using the search/replace function of Textadept.

**Reply by a1ess** [reply]
Update: I completely forgot that Textadept allows to pipe the selected text to shell commands. So, to achieve what I wanted, select text, Ctrl+|, then:

tr '\n' ','

**Comment by orbitalquark** [comment]
It's possible that there's a bug here. I will look into it more when I have some time. In the meantime, you can input the following into the Lua command entry:

  if buffer.selection_empty then buffer:set_text((buffer:get_text():gsub('\n',','))) else buffer:replace_sel((buffer:get_sel_text():gsub('\n',','))) end

If no text is selected, replace all newlines with ',' and if there is text selected, replace '\n' in only that text with ','

**Reply by a1ess** [reply]
Thank you so much for the code! This solution is better, as I can put it in my init.lua and associate it with a shortcut :)

**Comment by orbitalquark** [comment]
Oh nice, I didn't realize you were on Linux. Yes, shell commands to the rescue. Nicely done!

**Comment by orbitalquark** [comment]
That works, as does assigning a key to:

   function() textadept.editing.filter_through("tr '\n' ',') end

`textadept.editing.filter_through()` handles selections a bit better in my opinion[1].

[1]: https://orbitalquark.github.io/textadept/api.html#textadept.editing.filter_through

**Reply by a1ess** [reply]
Thank for this! I had been checking the API to see if there was a way to call shell commands from a function :)

**Comment by orbitalquark** [comment]
Okay, I've fixed the underlying bug via https://github.com/orbitalquark/textadept/commit/57d946f2eba8de5f7386b553558567b470900322

You can now "Replace All" with empty regex matches like '$', '^', etc.

**Reply by a1ess** [reply]
Hi,

I'm testing with Textadept v11.3 beta, but I still get the same behavior. If I try a Replace All as follows:

Find: $
Replace: ,
Regex: on 

with this sample:

abc
def
ghi

I get:

abc,
def,
ghi,

The comma is added after the last character, but the newline remains in place, as the text is still on three lines instead of becoming one line with the newlines replaced by commas.

I also tried this regex match, and found out that Textadept cannot match a whole line:

Find: ^.*$

Returns "No results found".

Strangely it works with: ^.*.$ and with ^.+$

I tried the search on Geany, which also uses Scintilla, and it matches a whole line with ^.$

Is the implementation of regexes in Textadept slightly different?

**Comment by orbitalquark** [comment]
In regex, '$' actually means match the position at the end-of-line, not the EOL itself, so the search and replace you are seeing is the expected behavior. Scintilla's built-in regex does not allow multi-line matches, so squashing multiple lines onto one with a regex is not possible, hence the need for Lua scripts or filter-through.

Finding '^.*$' seems to work for me given the 'abc\ndef\nghi' string you gave, both in 11.2 and 11.3 beta.

I don't know what regex Geany uses, but Textadept uses Scintilla's built-in C++11 regex capabilities.

Given the age of this discussion, I'm not sure if there's a regression here or not. Please let me know if you find one. My unit tests and manual tests are passing right now.

**Reply by a1ess** [reply]
> Scintilla's built-in regex does not allow multi-line matches, so squashing multiple lines onto one with a regex is not possible [...]

Ok, that's what I thought. Actually Geany supports replacement of \n only when the multi-line option is activated, so probably it implements its own method.

> Finding '^.*$' seems to work for me given the 'abc\ndef\nghi' string you gave, both in 11.2 and 11.3 beta.

I think I found the problem, I tried again using 11.3 beta with a clean profile. Try adding a blank line before, like this:

[BLANKLINE]
abc
def
ghi

Now place the cursor on the firs (blank) line and try the following searches with Regex ticked:

Search : $
Result: Match 1/5

It finds the first match and indicates correctly that there are five matches (I have a blank line also after "ghi"), but clicking on "Find Next" does not advance to the next line.

With the same set up, try:

Search: ^.*$
Result: Match 1/5

Same behavior as above, the cursor won't move from the first line. If you use instead: ^.*.$, the lines are matched and "Find  Next" works as expected. Of course it only matches the 3 lines with characters in them, as the last regex doesn't match empty lines.

So probably the problem is when matching blank lines. 

Hope it helps.

**Comment by orbitalquark** [comment]
Thanks, I see now. It's an issue of Find vs. Find Incremental. The latter advances the search, but the former does not. I'll look into this when I have some time.

**Comment by orbitalquark** [comment]
This should be fixed via https://github.com/orbitalquark/textadept/commit/f81316624d077360447fd4ba977b7cb22ee2061b
